

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Evolution &amp; phylogenetic reconstruction &#8212; Evolution of Language &amp; Music Computer Labs</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Evolution_&_phylogenetic_reconstruction';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Bioacoustics in Praat" href="Bioacoustics_in_Praat.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to the Evolamus Computer Labs
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bioacoustics_in_Praat.html">Bioacoustics in Praat</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Evolution &amp; phylogenetic reconstruction</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/clclab/elm-labs/blob/main/book/Evolution_&_phylogenetic_reconstruction.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/clclab/elm-labs" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/clclab/elm-labs/issues/new?title=Issue%20on%20page%20%2FEvolution_&_phylogenetic_reconstruction.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Evolution_&_phylogenetic_reconstruction.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Evolution & phylogenetic reconstruction</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#getting-started-with-r">0. Getting started with R</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#installing-packages">Installing packages</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-dataset">Downloading dataset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#auxiliary-functions">Auxiliary functions</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulated-evolution">1. Simulated evolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-population">1.1 Creating a population</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1-creating-a-genotype"><strong>Exercise 1: Creating a genotype</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-creating-the-population"><strong>Exercise 2: Creating the population</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-a-fitness-function">1.2 Defining a fitness function</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-compute-each-member-s-fitness-value"><strong>Exercise 3: Compute each member’s fitness value</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#implementing-selection">1.3 Implementing selection</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-4-a-new-population"><strong>Exercise 4: A new population</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-5-population-fitness-over-generations"><strong>Exercise 5: Population fitness over generations</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evolution-with-mutation">1.4 Evolution with mutation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-6-mu-tating-genomes"><strong>Exercise 6: Mu-tating genomes</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulated-evolution-with-ancestry">2. Simulated evolution with ancestry</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tracking-ancestry">2.1 Tracking ancestry</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-7-finding-the-ancestors"><strong>Exercise 7: Finding the ancestors</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-family-tree">2.2 Plotting the family tree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-8-ancestors-in-a-tree"><strong>Exercise 8: Ancestors in a tree</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-reconstruction-of-biological-data">3. Phylogenetic reconstruction of biological data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-9-animal-species-distances"><strong>Exercise 9: Animal species distances</strong></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-10-animal-species-phylogeny"><strong>Exercise 10: Animal species phylogeny</strong></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-trees-with-hierarchical-clustering">4. Phylogenetic trees with hierarchical clustering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#upgma">4.1 UPGMA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constructing-a-tree">4.2 Constructing a tree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-11-animal-species-phylogeny-with-upgma"><strong>Exercise 11: Animal species phylogeny, with UPGMA</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-12-animal-phylogenetic-tree-from-upgma"><strong>Exercise 12: Animal phylogenetic tree, from UPGMA</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-13-our-simulated-ancestors-in-a-tree-with-upgma"><strong>Exercise 13: Our simulated ancestors in a tree, with UPGMA</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-reconstruction-of-languages">5. Phylogenetic reconstruction of languages</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-14-languages-in-a-tree-with-upgma"><strong>Exercise 14: Languages in a tree, with UPGMA</strong></a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="evolution-phylogenetic-reconstruction">
<h1>Evolution &amp; phylogenetic reconstruction<a class="headerlink" href="#evolution-phylogenetic-reconstruction" title="Permalink to this heading">#</a></h1>
<p><em>Written by Bastiaan van der Weij and Dieuwke Hupkes.<br />
Updated by Bas Cornelissen, Peter Dekker and Marianne de Heer Kloots.</em></p>
<p>In this computer lab you will experiment with simulated evolution and phylogenetic reconstruction: the reconstruction of the evolutionary history of species, simulated genes, and even language and music.</p>
<p>The lab contains <font color="orange"><b>Code</b></font> questions, where you need to enter some code to compute or simulate something yourself, and <font color="green"><b>Think</b></font> questions, which check your understanding of what is happening.</p>
<p>Some notes on using this notebook:</p>
<ul class="simple">
<li><p>You should run all the code cells in this notebook, either by clicking the ‘play button’ that appears once you hover over them (in Google Colab) or by pressing Shift+Enter from inside each cell. Make sure you don’t miss any! Once a cell has completed running, a number will appear between the [square brackets] to the left of the cell.</p></li>
<li><p>You may run a cell multiple times (for example to check how the output changes).</p></li>
<li><p>In principle, you could also run a previous cell again after you’ve run the cells below it. However, this might be a bad idea, and cause you trouble once you try to run new code further down the line again. If this happens and you run into strange errors, you can always try to restart the notebook (Ctrl+M or <em>Runtime &gt; Restart runtime</em> in the menu) and carefully run each cell from top to bottom again.</p></li>
<li><p>To see an overview of the entire computer lab, and easily navigate between sections, open the <em>Table of contents</em> tab to the left (first icon in the navigation bar).</p></li>
</ul>
<section id="getting-started-with-r">
<h2>0. Getting started with R<a class="headerlink" href="#getting-started-with-r" title="Permalink to this heading">#</a></h2>
<p>We assume you have some basic understanding of
programming and R. You will not have to write a lot of code yourself,
but you should be able to read and run the scripts that we provide you
with. If you have never seen any R code before, it might be useful for
you to work through a tutorial, such as the first three pages of <a class="reference external" href="http://projects.illc.uva.nl/LaCo/CLAS/clc13/assignments/deboer13rtutorial.pdf">this
tutorial by Bart de
Boer</a>.
Of course, you can always ask us for help if you are stuck on a
particular piece of code!</p>
<section id="installing-packages">
<h3>Installing packages<a class="headerlink" href="#installing-packages" title="Permalink to this heading">#</a></h3>
<p>This cell installs and loads some external packages that we will need throughout the computer lab. It may take a few minutes to run!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># list of packages to install</span>
<span class="n">to_install</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;stringr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ape&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;phangorn&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;plotly&quot;</span><span class="p">)</span>

<span class="c1"># create writable library for the packages</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;library&quot;</span><span class="p">)){</span>
<span class="w">    </span><span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;library&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># install packages to the library</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">all</span><span class="p">(</span><span class="nf">file.exists</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&#39;library/&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">to_install</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)))){</span>
<span class="w">    </span><span class="n">to_install</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">to_install</span><span class="p">[</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&#39;library/&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">to_install</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))]</span>
<span class="w">    </span><span class="nf">install.packages</span><span class="p">(</span><span class="n">to_install</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;library/&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># tell R where the library is</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="s">&#39;library/&#39;</span><span class="p">)</span>

<span class="c1"># load packages from the library</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">phangorn</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">plotly</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="downloading-dataset">
<h3>Downloading dataset<a class="headerlink" href="#downloading-dataset" title="Permalink to this heading">#</a></h3>
<p>This cell downloads a dataset that we have preprocessed for you, to use in <span class="xref myst">section 5</span> of this computer lab.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;data&quot;</span><span class="p">)){</span>
<span class="w">    </span><span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;data&quot;</span><span class="p">)</span>
<span class="p">}</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">file.exists</span><span class="p">(</span><span class="s">&quot;data/language_data.Rdata&quot;</span><span class="p">)){</span>
<span class="w">    </span><span class="nf">download.file</span><span class="p">(</span><span class="s">&quot;https://github.com/clclab/elm-labs/raw/main/book/language_data.Rdata&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">              </span><span class="s">&quot;data/language_data.Rdata&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s">&quot;wb&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="auxiliary-functions">
<h3>Auxiliary functions<a class="headerlink" href="#auxiliary-functions" title="Permalink to this heading">#</a></h3>
<p>The code below defines some of our own functions and variables that we will use throughout this computer lab. Run the cell so they are loaded; afterwards, you can collapse the cell and directly <span class="xref myst">skip to the first section</span> for now. But if you want to examine these functions more closely later on, you can go back to this cell (and expand it) to do so.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># alphabet</span>
<span class="n">alphabet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">,</span><span class="s">&quot;G&quot;</span><span class="p">,</span><span class="s">&quot;U&quot;</span><span class="p">,</span><span class="s">&quot;C&quot;</span><span class="p">)</span>

<span class="c1"># target matrices if fixed target is used</span>
<span class="n">target_R</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="m">3</span><span class="p">,</span><span class="m">3</span><span class="p">)</span>
<span class="n">target_S</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="m">3</span><span class="p">,</span><span class="m">3</span><span class="p">)</span>
<span class="n">target_agent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">R</span><span class="o">=</span><span class="n">target_R</span><span class="p">,</span><span class="w"> </span><span class="n">S</span><span class="o">=</span><span class="n">target_S</span><span class="p">)</span>

<span class="n">CAC_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># fitness function based on number of</span>
<span class="w">  </span><span class="c1"># occurences of substring CAC</span>
<span class="w">  </span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">population_size</span><span class="p">)</span>
<span class="w">  </span><span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">population_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">population_member</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">,],</span><span class="w"> </span><span class="n">collapse</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="w">    </span><span class="n">fitness</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">str_count</span><span class="p">(</span><span class="n">population_member</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CAC&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">fitness</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Mutation function</span>

<span class="n">mutation_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="m">3</span><span class="p">,</span><span class="m">16</span><span class="p">),</span><span class="m">4</span><span class="p">,</span><span class="m">4</span><span class="p">)</span>
<span class="nf">diag</span><span class="p">(</span><span class="n">mutation_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>

<span class="n">mutate_population_fast</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># choose number of elements to change</span>
<span class="w">  </span><span class="n">N_change</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbinom</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="o">*</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="o">=</span><span class="n">mu</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># generate indices of elements to change</span>
<span class="w">  </span><span class="n">indices</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="n">genome_size</span><span class="o">*</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="n">N_change</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># change elements</span>
<span class="w">  </span><span class="nf">for </span><span class="p">(</span><span class="n">index</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">indices</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">row</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ceiling</span><span class="p">(</span><span class="n">index</span><span class="o">/</span><span class="n">genome_size</span><span class="p">)</span>
<span class="w">    </span><span class="n">column</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="o">%%</span><span class="w"> </span><span class="n">genome_size</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span>
<span class="w">    </span><span class="n">distr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mutation_matrix</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">population</span><span class="p">[</span><span class="n">row</span><span class="p">,</span><span class="w"> </span><span class="n">column</span><span class="p">],</span><span class="w"> </span><span class="n">alphabet</span><span class="p">),]</span>
<span class="w">    </span><span class="n">population</span><span class="p">[</span><span class="n">row</span><span class="p">,</span><span class="w"> </span><span class="n">column</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="n">alphabet</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="o">=</span><span class="n">distr</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">10</span><span class="p">))</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;A&quot;</span><span class="p">,</span><span class="s">&quot;G&quot;</span><span class="p">,</span><span class="s">&quot;C&quot;</span><span class="p">,</span><span class="s">&quot;U&quot;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">collapse</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="p">}</span>


<span class="c1">## Auxiliary functions for phylogenetic trees</span>

<span class="n">compute_diversity</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">population</span><span class="p">)))</span>
<span class="p">}</span>

<span class="c1"># construct a tree from a population using a parent matrix</span>
<span class="n">reconstruct_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">D</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span><span class="w">    </span><span class="c1"># maximum depth tree</span>
<span class="w">  </span><span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ncol</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span><span class="w">    </span><span class="c1"># population size</span>
<span class="w">  </span><span class="n">N_cur</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">N</span><span class="w">                  </span><span class="c1"># cur population size</span>
<span class="w">  </span><span class="n">cur_depth</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">              </span><span class="c1"># cur depth tree</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">children_indices</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="m">1</span><span class="p">)</span>
<span class="w">  </span><span class="n">children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="nf">while </span><span class="p">(</span><span class="n">N_cur</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="n">cur_depth</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">D</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># while (cur_depth &lt; D) {</span>
<span class="w">    </span><span class="c1"># create an empty list of population size to group children</span>
<span class="w">    </span><span class="c1"># under the right parent (index)</span>
<span class="w">    </span><span class="n">parents</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">vector</span><span class="p">(</span><span class="s">&quot;list&quot;</span><span class="p">,</span><span class="n">N</span><span class="p">)</span>
<span class="w">    </span><span class="n">has_children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">N</span><span class="p">)</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># loop over the children by looping over their indices</span>
<span class="w">    </span><span class="n">child_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">child</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">children_indices</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">parent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parent_matrix</span><span class="p">[</span><span class="n">D</span><span class="o">-</span><span class="n">cur_depth</span><span class="p">,</span><span class="n">child</span><span class="p">]</span><span class="w">  </span><span class="c1"># find index of parent in prev generation</span>
<span class="w">      </span><span class="n">child</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">children</span><span class="p">[[</span><span class="n">child_index</span><span class="p">]]</span><span class="w">            </span><span class="c1"># get child from array</span>
<span class="w">      </span><span class="n">cur_children_parent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]]</span><span class="w">    </span><span class="c1"># check if parent already is assigned children</span>
<span class="w">      </span><span class="n">cur_number_children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">has_children</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span><span class="w"> </span><span class="c1"># check number of children for parent</span>
<span class="w">      </span>
<span class="w">      </span><span class="c1"># assign child to parent</span>
<span class="w">      </span><span class="nf">if </span><span class="p">(</span><span class="n">cur_number_children</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="n">cur_depth</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">        </span><span class="c1"># first child, child is already a tree</span>
<span class="w">        </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">children</span><span class="p">[[</span><span class="n">child_index</span><span class="p">]]</span>
<span class="w">      </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="nf">if </span><span class="p">(</span><span class="n">cur_number_children</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">                  </span><span class="c1"># first child, child is leaf node</span>
<span class="w">        </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">children</span><span class="p">[</span><span class="n">child_index</span><span class="p">])</span>
<span class="w">      </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="nf">if </span><span class="p">(</span><span class="n">cur_depth</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">                            </span><span class="c1"># not first child, child is leaf node</span>
<span class="w">        </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]][[</span><span class="n">cur_number_children</span><span class="m">+1</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">child</span>
<span class="w">      </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="nf">if </span><span class="p">(</span><span class="n">cur_number_children</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">                  </span><span class="c1"># second child, child is already a tree, children is not yet a list</span>
<span class="w">        </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">cur_children_parent</span><span class="p">,</span><span class="w"> </span><span class="n">child</span><span class="p">)</span>
<span class="w">      </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="p">{</span><span class="w">                                                </span><span class="c1"># third or more child</span>
<span class="w">        </span><span class="n">parents</span><span class="p">[[</span><span class="n">parent</span><span class="p">]][[</span><span class="n">cur_number_children</span><span class="m">+1</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">child</span>
<span class="w">      </span><span class="p">}</span>
<span class="w">      </span>
<span class="w">      </span><span class="n">has_children</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">has_children</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="w">        </span><span class="c1"># increase children for parent</span>
<span class="w">      </span>
<span class="w">      </span><span class="n">child_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">child_index</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="n">nulls</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="n">parents</span><span class="p">,</span><span class="w"> </span><span class="n">FUN</span><span class="o">=</span><span class="n">is.null</span><span class="p">)</span><span class="w">       </span><span class="c1"># check which prev population members didn&#39;t have children</span>
<span class="w">    </span><span class="n">children_indices</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="m">1</span><span class="p">)[</span><span class="o">!</span><span class="n">nulls</span><span class="p">]</span><span class="w">   </span><span class="c1"># get the indices of the parents, set children to them</span>
<span class="w">    </span><span class="n">children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parents</span><span class="p">[</span><span class="o">!</span><span class="n">nulls</span><span class="p">]</span>
<span class="w">    </span><span class="n">N_cur</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">children_indices</span><span class="p">)</span>
<span class="w">    </span><span class="n">cur_depth</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cur_depth</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">if </span><span class="p">(</span><span class="n">N_cur</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">print</span><span class="p">(</span><span class="s">&quot;Disconnected tree&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span><span class="w"> </span><span class="n">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">children</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">}</span>

<span class="c1"># print a nested_list in newick-format</span>
<span class="n">print_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">nested_list</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="nf">capture.output</span><span class="p">(</span><span class="nf">dput</span><span class="p">(</span><span class="n">nested_list</span><span class="p">)),</span><span class="n">collapse</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;[\r\n]&quot;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">str_repr</span><span class="p">)</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&#39;\&quot;&#39;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">str_repr</span><span class="p">)</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&#39;list&#39;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">str_repr</span><span class="p">)</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="n">str_repr</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;;&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">str_repr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="n">str_repr</span><span class="p">)</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">str_repr</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># plot tree</span>
<span class="n">plot_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="nf">print_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">),</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s">&quot;tree.nwk&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">tree_sim</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.tree</span><span class="p">(</span><span class="s">&quot;tree.nwk&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="nf">plot</span><span class="p">(</span><span class="n">tree_sim</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">0.45</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># print connections in the parent matrix</span>
<span class="n">print_parent_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">);</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ncol</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span>
<span class="w">  </span><span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="n">height</span><span class="p">)</span><span class="o">*</span><span class="n">width</span>
<span class="w">  </span><span class="n">x_from</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">width</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="n">times</span><span class="o">=</span><span class="n">height</span><span class="p">)</span>
<span class="w">  </span><span class="n">y_from</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">height</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="n">each</span><span class="o">=</span><span class="n">width</span><span class="p">)</span>
<span class="w">  </span><span class="n">y_to</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="n">height</span><span class="m">+1</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="n">each</span><span class="o">=</span><span class="n">width</span><span class="p">)</span>
<span class="w">  </span><span class="n">x_to</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">N</span><span class="p">)</span>
<span class="w">  </span><span class="n">counter</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="p">;</span>
<span class="w">  </span><span class="nf">for </span><span class="p">(</span><span class="n">population</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">height</span><span class="o">:</span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">individual</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">width</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">x_to</span><span class="p">[</span><span class="n">counter</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">parent_matrix</span><span class="p">[</span><span class="n">population</span><span class="p">,</span><span class="n">individual</span><span class="p">]</span><span class="w"> </span>
<span class="w">      </span><span class="n">counter</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">counter</span><span class="m">+1</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="w">  </span><span class="nf">plot</span><span class="p">(</span><span class="n">width</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="m">+2</span><span class="p">,</span><span class="w"> </span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">width</span><span class="p">),</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">height</span><span class="m">+2</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">yaxt</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">xaxt</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;population member&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;generation&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="nf">axis</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">at</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">))</span>
<span class="w">  </span><span class="nf">segments</span><span class="p">(</span><span class="n">x_from</span><span class="p">,</span><span class="w"> </span><span class="n">y_from</span><span class="p">,</span><span class="w"> </span><span class="n">x_to</span><span class="p">,</span><span class="w"> </span><span class="n">y_to</span><span class="p">)</span>
<span class="w">  </span><span class="nf">points</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">x_from</span><span class="p">,</span><span class="w"> </span><span class="n">x_to</span><span class="p">),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">y_from</span><span class="p">,</span><span class="w"> </span><span class="n">y_to</span><span class="p">),</span><span class="w"> </span><span class="n">pch</span><span class="o">=</span><span class="m">19</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># compute distance matrix</span>
<span class="n">compute_distance_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="w">  </span><span class="n">dm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">N</span><span class="o">*</span><span class="n">N</span><span class="p">),</span><span class="n">N</span><span class="p">,</span><span class="n">N</span><span class="p">)</span>
<span class="w">  </span><span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="p">(</span><span class="n">N</span><span class="m">-1</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">j</span><span class="w"> </span><span class="nf">in </span><span class="p">(</span><span class="n">i</span><span class="m">+1</span><span class="p">)</span><span class="o">:</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">agent1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">,]</span>
<span class="w">      </span><span class="n">agent2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">population</span><span class="p">[</span><span class="n">j</span><span class="p">,]</span>
<span class="w">      </span><span class="n">dm</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_distance</span><span class="p">(</span><span class="n">agent1</span><span class="p">,</span><span class="w"> </span><span class="n">agent2</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">colnames</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">N</span><span class="p">))</span>
<span class="w">  </span><span class="nf">rownames</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">N</span><span class="p">))</span>
<span class="w">  </span><span class="n">dm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.dist</span><span class="p">(</span><span class="n">dm</span><span class="p">,</span><span class="n">diag</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">upper</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># compute normalised hamming distance between two strings</span>
<span class="n">compute_distance</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">agent1</span><span class="p">,</span><span class="w"> </span><span class="n">agent2</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">l</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">agent1</span><span class="p">)</span>
<span class="w">  </span><span class="n">helper</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">l</span><span class="p">)</span>
<span class="w">  </span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">helper</span><span class="p">[</span><span class="o">!</span><span class="n">agent1</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">agent2</span><span class="p">])</span><span class="o">/</span><span class="n">l</span>
<span class="w">  </span><span class="nf">return</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="simulated-evolution">
<h2>1. Simulated evolution<a class="headerlink" href="#simulated-evolution" title="Permalink to this heading">#</a></h2>
<p>The goals of this section are:</p>
<ul class="simple">
<li><p>to better understand the concepts of genotype, genotype space,
fitness, fitness landscape, selection, mutation, selection-mutation
balance, frequency dependent selection;</p></li>
<li><p>to see how these concepts can be formalised in a computer program;</p></li>
<li><p>to appreciate both the power and the limits of natural selection.</p></li>
</ul>
<section id="creating-a-population">
<h3>1.1 Creating a population<a class="headerlink" href="#creating-a-population" title="Permalink to this heading">#</a></h3>
<p>In the first part of this computer lab, we will use R to simulate the evolution of (DNA) strings under a particular fitness function. We will represent DNA strings (the genotype) as a sequence of the letters ‘A’, ‘G’, ‘C’ and ‘U’. In R, you can generate a random sequence of 10 of these letters using the following command:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>To store the output in a variable (for instance x), you type:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>You can then view the contents of a particular variable by printing it.</p>
<p>Let’s try this, by running the cell below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [1] &quot;C&quot; &quot;C&quot; &quot;A&quot; &quot;C&quot; &quot;C&quot; &quot;U&quot; &quot;G&quot; &quot;U&quot; &quot;G&quot; &quot;U&quot;
</pre></div>
</div>
</div>
</div>
<section id="exercise-1-creating-a-genotype">
<h4><strong>Exercise 1: Creating a genotype</strong><a class="headerlink" href="#exercise-1-creating-a-genotype" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p>First, use the cell above to generate a few random sequences of length 10 containing the characters ‘A’, ‘G’, ‘C’ and ‘U’, and confirm that it does what we want it to do.</p></li>
<li><p><font color="orange"><b>Code:</b></font> In the code cell below, generate a random sequence of length 50 containing the characters ‘A’, ‘G’, ‘C’ and ‘U’.</p></li>
<li><p><font color="green"><b>Think:</b></font> The set of all possible sequences is called the <em>genotype space</em>. How big is this space? I.e., how many genotype strings are possible with our representation?</p></li>
</ul>
<p>(Note that R ignores everything that follows the character #. In programming terms, these texts are called <em>comments</em>.)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="w"> </span><span class="c1">### YOUR CODE HERE ###</span>
<span class="w"> </span>
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
</section>
<section id="exercise-2-creating-the-population">
<h4><strong>Exercise 2: Creating the population</strong><a class="headerlink" href="#exercise-2-creating-the-population" title="Permalink to this heading">#</a></h4>
<p>Now, let’s create a <em>population</em> of DNA strings. To do this, we will make 100 genotype strings. We will store our population in a matrix (the <em>population matrix</em>), where each member of our population is represented as a row of the matrix.</p>
<p>The command <code class="docutils literal notranslate"><span class="pre">matrix(x,</span> <span class="pre">height,</span> <span class="pre">width)</span></code> command transforms a vector <code class="docutils literal notranslate"><span class="pre">x</span></code> into a matrix with height <code class="docutils literal notranslate"><span class="pre">height</span></code> and width <code class="docutils literal notranslate"><span class="pre">width</span></code>.</p>
<p>Let’s start with creating a matrix filled with zero’s that we can later fill:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> So what do we find in column 30 of our population matrix, and what does this number mean?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
<p><code class="docutils literal notranslate"><span class="pre">population[,j]</span></code> accesses the <code class="docutils literal notranslate"><span class="pre">j</span></code>th column of the matrix <code class="docutils literal notranslate"><span class="pre">population</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">population[i,]</span></code> accesses the <code class="docutils literal notranslate"><span class="pre">i</span></code>th <em>row</em> of the matrix <code class="docutils literal notranslate"><span class="pre">population</span></code>, which in our case corresponds with the <code class="docutils literal notranslate"><span class="pre">i</span></code>th member of our population.</p>
<p>We fill our matrix by generating 100 population members in a for-loop and filling the matrix with them:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">population_size</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">population_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">,]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span>
<span class="w">      </span><span class="n">size</span><span class="o">=</span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="defining-a-fitness-function">
<h3>1.2 Defining a fitness function<a class="headerlink" href="#defining-a-fitness-function" title="Permalink to this heading">#</a></h3>
<p>Now we need to define a fitness function that computes the fitness of
the individual members of our population. Imagine, for instance, that
the string ‘CAC’ codes for some very useful aminoacid. The more CAC’s in
the genome, the higher the expected number of offspring. In our
simulation of evolution, let’s define the fitness as the number of times
the substring ‘CAC’ appears in the genotype string (without overlap, so
the string ‘CACAC’ contains <em>one</em> copy of ‘CAC’).</p>
<p>To keep track of the fitness of <em>all</em> members in our population (which
are represented as rows in the population matrix), we create a <em>vector</em>
where each element of the vector represents the fitness value for one
member of the population.</p>
<section id="exercise-3-compute-each-member-s-fitness-value">
<h4><strong>Exercise 3: Compute each member’s fitness value</strong><a class="headerlink" href="#exercise-3-compute-each-member-s-fitness-value" title="Permalink to this heading">#</a></h4>
<p>We generate an empty vector to store the fitness values, and call it <code class="docutils literal notranslate"><span class="pre">fitness</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">population_size</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now, we use a for-loop to fill the vector (created by the code above) with the fitness values:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># loop over population size</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">population_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">      </span>
<span class="w">  </span><span class="c1"># generate string representation</span>
<span class="w">  </span><span class="n">member</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">,],</span><span class="w"> </span><span class="n">collapse</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">)</span><span class="w"> </span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># compute fitness member   </span>
<span class="w">  </span><span class="n">fitness_member</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">str_count</span><span class="p">(</span><span class="n">member</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CAC&quot;</span><span class="p">)</span><span class="w">   </span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># store in fitness vector</span>
<span class="w">  </span><span class="n">fitness</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fitness_member</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> What is the highest possible fitness a member of this population can have?</p></li>
<li><p><font color="orange"><b>Code:</b></font> Compute the mean fitness of your population, by using <code class="docutils literal notranslate"><span class="pre">mean(fitness)</span></code> in the code cell below and storing the result to a new variable <code class="docutils literal notranslate"><span class="pre">av_fitness</span></code>. What is the average fitness of your population?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="implementing-selection">
<h3>1.3 Implementing selection<a class="headerlink" href="#implementing-selection" title="Permalink to this heading">#</a></h3>
<p>Now we will generate the next generation. We assume that each member of the next generation inherits the genome of one of the members of the previous generation. The probability of inheriting each genome is proportional to the genome’s fitness: a child is most likely to inherit the genome of the fittest member of the previous population. This simulates selection.</p>
<section id="exercise-4-a-new-population">
<h4><strong>Exercise 4: A new population</strong><a class="headerlink" href="#exercise-4-a-new-population" title="Permalink to this heading">#</a></h4>
<p>We generate 100 new children, using the built-in function <code class="docutils literal notranslate"><span class="pre">sample</span></code> (the same one we used before). We first draw 100 random numbers between 1 and 100 (repetitions possible), and store them in the variable <code class="docutils literal notranslate"><span class="pre">indices</span></code>. The likelihood of being drawn depends on a member’s fitness: if population member 2 has a very high fitness, it will have a very high chance of being drawn. Then we use the drawn numbers to create a new population of the members corresponding to the numbers.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">indices</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span>
<span class="w">                </span><span class="n">prob</span><span class="o">=</span><span class="n">fitness</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">fitness</span><span class="p">))</span>
<span class="n">new_population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">population</span><span class="p">[</span><span class="n">indices</span><span class="p">,]</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> If one population member has fitness 10 and all the other population members have fitness 1, what is the probability that a child will inherit its genome from this one population member? What do you expect to happen with the population?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
</section>
<section id="exercise-5-population-fitness-over-generations">
<h4><strong>Exercise 5: Population fitness over generations</strong><a class="headerlink" href="#exercise-5-population-fitness-over-generations" title="Permalink to this heading">#</a></h4>
<p>To simulate the evolution of the population, we want to repeat this process several times and plot the average fitness over time. The cell below defines the functions <code class="docutils literal notranslate"><span class="pre">generate_population</span></code> and <code class="docutils literal notranslate"><span class="pre">simulate_evolution</span></code>, which allow us to do this easily.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to generate new population</span>
<span class="n">generate_population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="nf">function</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># generate initial population with random strings</span>
<span class="w">    </span><span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">population_size</span><span class="o">*</span><span class="n">genome_size</span><span class="p">),</span><span class="w"> </span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">population_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">new_member</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="n">genome_size</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">        </span><span class="n">population</span><span class="p">[</span><span class="n">i</span><span class="p">,]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">new_member</span>
<span class="w">    </span><span class="p">}</span>

<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Function to simulate evolution</span>
<span class="n">simulate_evolution</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>

<span class="w">    </span><span class="c1"># compute fitness</span>
<span class="w">    </span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CAC_count</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

<span class="w">    </span><span class="n">av_fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">simulation_length</span><span class="p">)</span>
<span class="w">    </span><span class="n">av_fitness</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mean</span><span class="p">(</span><span class="n">fitness</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># simulate evolution</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">j</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="n">simulation_length</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>

<span class="w">        </span><span class="c1"># generate children by sampling from the population according to fitness</span>
<span class="w">        </span><span class="n">population_children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">population</span><span class="p">[</span><span class="nf">sample</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="o">=</span><span class="n">fitness</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">fitness</span><span class="p">)),]</span>

<span class="w">        </span><span class="c1"># mutate children</span>
<span class="w">        </span><span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mutate_population_fast</span><span class="p">(</span><span class="n">population_children</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># recompute fitness</span>
<span class="w">        </span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CAC_count</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># add to list with average fitness</span>
<span class="w">        </span><span class="n">av_fitness</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mean</span><span class="p">(</span><span class="n">fitness</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>

<span class="w">    </span><span class="c1"># plot average fitness</span>
<span class="w">    </span><span class="n">ymax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">av_fitness</span><span class="p">[</span><span class="n">simulation_length</span><span class="p">]</span><span class="m">+2</span>
<span class="w">    </span><span class="n">generation</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">simulation_length</span><span class="p">,</span><span class="m">1</span><span class="p">)</span>
<span class="w">    </span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="w">    </span><span class="nf">plot</span><span class="p">(</span><span class="n">generation</span><span class="p">,</span><span class="w"> </span><span class="n">av_fitness</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s">&quot;l&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ann</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">ymax</span><span class="p">))</span>
<span class="w">    </span><span class="nf">title</span><span class="p">(</span><span class="n">main</span><span class="o">=</span><span class="s">&quot;Average population fitness&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;Generation&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Fitness&quot;</span><span class="p">)</span>

<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>Now, we can run a simulation with our desired settings. Try it out with the cell below (you can ignore the parameter <code class="docutils literal notranslate"><span class="pre">mu</span></code> for now, we will get to it very soon).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># parameters of the simulation</span>
<span class="n">population_size</span><span class="w">	</span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span>
<span class="n">genome_size</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">50</span>
<span class="n">simulation_length</span><span class="w">	</span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span>
<span class="n">mu</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>

<span class="c1"># Generate population and simulate evolution</span>
<span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">generate_population</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span>
<span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">simulate_evolution</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/fd8d9d3cf294877698fe1e8d4030da1237f881f48b741d73c4413a98eb938a69.png" src="_images/fd8d9d3cf294877698fe1e8d4030da1237f881f48b741d73c4413a98eb938a69.png" />
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> You will notice the fitness stops increasing quite early in the simulation. Why is this?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
</section>
</section>
<section id="evolution-with-mutation">
<h3>1.4 Evolution with mutation<a class="headerlink" href="#evolution-with-mutation" title="Permalink to this heading">#</a></h3>
<p>In the previous simulation, we looked at selection <em>without</em> mutation. Let’s now look at the case where every child’s nucleotide has a probability <span class="math notranslate nohighlight">\(\mu\)</span> to change into a random other nucleotide.</p>
<section id="exercise-6-mu-tating-genomes">
<h4><strong>Exercise 6: Mu-tating genomes</strong><a class="headerlink" href="#exercise-6-mu-tating-genomes" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> If <span class="math notranslate nohighlight">\(\mu = 0.01\)</span>, what is the chance that no changes occur in a genome? What is the chance that no changes occur in an entire population? And if <span class="math notranslate nohighlight">\(\mu = 0.001\)</span> ?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
<ul class="simple">
<li><p><font color="orange"><b>Code:</b></font> In the first code cell below, copy-paste the code from the last code cell above, but change the mutation level to <span class="math notranslate nohighlight">\(\mu = 0.001\)</span>. Also adapt the length of the simulation to 1000 generations. Then run the cell to see the results.</p></li>
<li><p><font color="orange"><b>Code:</b></font> In the second code cell below, repeat the simulation again, but now with <span class="math notranslate nohighlight">\(\mu = 0.01\)</span>, and compare the two plots. This shows the mutation-selection balance.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> In the simulation with relatively high mutation rate, why does the fitness stop increasing at a slightly lower level?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
</section>
</section>
</section>
<section id="simulated-evolution-with-ancestry">
<h2>2. Simulated evolution with ancestry<a class="headerlink" href="#simulated-evolution-with-ancestry" title="Permalink to this heading">#</a></h2>
<p>In the previous section we implemented computer simulations of evolution
in R. In this section, we extend this simulation, and think about the
patterns of genetic variation that the evolutionary process leaves in a
population. We look at methods for <em>phylogenetic tree reconstruction</em>.
These methods use the genetic variation in the most recent (current)
generation to reconstruct the evolutionary history of a population or a
set of species.</p>
<p>We use a simple <em>clustering algorithm</em> for
phylogenetic tree reconstruction. An algorithm is a description of a series of steps to do arrive at a certain end result or perform a calculation. In this case, the computer performs the steps for us.</p>
<p>The goal of this section is to learn about the possibilities and
difficulties that are involved in phylogenetic tree reconstruction.</p>
<section id="tracking-ancestry">
<h3>2.1 Tracking ancestry<a class="headerlink" href="#tracking-ancestry" title="Permalink to this heading">#</a></h3>
<p>So far, we have simulated the evolution of strings of symbols, and looked at the effect of different fitness functions. Now we will repeat this simulation, but during the evolutionary process we will keep track of ancestry, so that we can reconstruct family trees of different individuals. We will start with a very simple simulation.</p>
<p>The functions <code class="docutils literal notranslate"><span class="pre">generate_population</span></code> and <code class="docutils literal notranslate"><span class="pre">simulate_evolution_ancestry</span></code> in the code cell below allow us to run the same simulation as above, but now also generating a matrix <code class="docutils literal notranslate"><span class="pre">parent_matrix</span></code>. This matrix specifies the parent of each member in each generation (where the parent is the individual of the previous generation whose genetic material was inherited).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to generate new population</span>
<span class="n">generate_population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="nf">function</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># generate initial population with random strings</span>
<span class="w">    </span><span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">,</span><span class="s">&#39;U&#39;</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="n">genome_size</span><span class="o">*</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span>
<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Function to simulate evolution with ancestry</span>
<span class="n">simulate_evolution_ancestry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">population</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>

<span class="w">    </span><span class="c1"># initialise</span>
<span class="w">    </span><span class="n">back_pointers</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">population_size</span><span class="o">*</span><span class="p">(</span><span class="n">simulation_length</span><span class="m">-1</span><span class="p">)),</span><span class="w"> </span><span class="n">simulation_length</span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="n">population_size</span><span class="p">)</span><span class="w"> </span>

<span class="w">    </span><span class="c1"># compute fitness and population diversity</span>
<span class="w">    </span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CAC_count</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

<span class="w">    </span><span class="n">av_fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">simulation_length</span><span class="p">)</span>
<span class="w">    </span><span class="n">av_fitness</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mean</span><span class="p">(</span><span class="n">fitness</span><span class="p">)</span>

<span class="w">    </span><span class="n">diversity</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">simulation_length</span><span class="p">)</span>
<span class="w">    </span><span class="n">diversity</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_diversity</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

<span class="w">    </span><span class="c1"># simulate evolution</span>
<span class="w">    </span><span class="nf">for </span><span class="p">(</span><span class="n">j</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="n">simulation_length</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>

<span class="w">        </span><span class="c1"># generate children. </span>
<span class="w">        </span><span class="n">indices_children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="o">=</span><span class="n">fitness</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">fitness</span><span class="p">))</span>
<span class="w">        </span><span class="n">population_children</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">population</span><span class="p">[</span><span class="n">indices_children</span><span class="p">,]</span>
<span class="w">        </span><span class="n">back_pointers</span><span class="p">[</span><span class="n">j</span><span class="m">-1</span><span class="p">,]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">indices_children</span>

<span class="w">        </span><span class="c1"># mutation childrn</span>
<span class="w">        </span><span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mutate_population_fast</span><span class="p">(</span><span class="n">population_children</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># recompute fitness</span>
<span class="w">        </span><span class="n">fitness</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CAC_count</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>

<span class="w">        </span><span class="c1"># store current fitness and diversity</span>
<span class="w">        </span><span class="n">av_fitness</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mean</span><span class="p">(</span><span class="n">fitness</span><span class="p">)</span>
<span class="w">        </span><span class="n">diversity</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_diversity</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>

<span class="w">    </span><span class="c1"># plot average fitness and population diversity</span>
<span class="w">    </span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span>
<span class="w">    </span><span class="n">ymax</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">av_fitness</span><span class="p">[</span><span class="n">simulation_length</span><span class="p">]</span><span class="m">+2</span>
<span class="w">    </span><span class="n">generation</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="n">simulation_length</span><span class="p">,</span><span class="m">1</span><span class="p">)</span>
<span class="w">    </span><span class="nf">plot</span><span class="p">(</span><span class="n">generation</span><span class="p">,</span><span class="w"> </span><span class="n">av_fitness</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s">&quot;l&quot;</span><span class="p">,</span><span class="n">ann</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">ymax</span><span class="p">))</span>
<span class="w">    </span><span class="nf">title</span><span class="p">(</span><span class="n">main</span><span class="o">=</span><span class="s">&quot;Average population fitness&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;Generation&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Fitness&quot;</span><span class="p">)</span>
<span class="w">    </span><span class="nf">plot</span><span class="p">(</span><span class="n">generation</span><span class="p">,</span><span class="w"> </span><span class="n">diversity</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s">&quot;l&quot;</span><span class="p">,</span><span class="n">ann</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="n">population_size</span><span class="p">))</span>
<span class="w">    </span><span class="nf">title</span><span class="p">(</span><span class="n">main</span><span class="o">=</span><span class="s">&quot;Population diversity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;Generation&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;Number of distinct phenotypes&quot;</span><span class="p">)</span>

<span class="w">    </span><span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="n">population</span><span class="p">,</span><span class="n">parent_matrix</span><span class="o">=</span><span class="n">back_pointers</span><span class="p">)</span>
<span class="w">    </span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">))</span>
<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>With the code cell below, we can run the simulation with our desired settings. At the end of the simulation, we see a plot with the average population fitness and a plot with the diversity of the population over generations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># parameters of simulation</span>
<span class="n">population_size</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span>
<span class="n">genome_size</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">50</span>
<span class="n">simulation_length</span><span class="w">	</span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span>
<span class="n">mu</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.005</span>

<span class="c1"># Generate population and simulate evolution</span>
<span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">generate_population</span><span class="p">(</span><span class="n">population_size</span><span class="p">,</span><span class="w"> </span><span class="n">genome_size</span><span class="p">)</span>
<span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">simulate_evolution_ancestry</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="n">population</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">population</span>
<span class="n">parent_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">parent_matrix</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/c637f0ae3595477feeb4fcbdd535da98d25c5b783ca38a1f262c924a2590f5da.png" src="_images/c637f0ae3595477feeb4fcbdd535da98d25c5b783ca38a1f262c924a2590f5da.png" />
</div>
</div>
<section id="exercise-7-finding-the-ancestors">
<h4><strong>Exercise 7: Finding the ancestors</strong><a class="headerlink" href="#exercise-7-finding-the-ancestors" title="Permalink to this heading">#</a></h4>
<p>Visualise the parent matrix by running the code cell below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">print_parent_matrix</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/896279b5d52c82008aa6e82c3870322bc12341ce839626346debb94f01cf80f0.png" src="_images/896279b5d52c82008aa6e82c3870322bc12341ce839626346debb94f01cf80f0.png" />
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> Where in this plot can you find the first generation?</p></li>
<li><p><font color="green"><b>Think:</b></font> Follow some paths up and down. Why do downward paths often end in dead ends, whereas upward paths always go all the way up?</p></li>
</ul>
<blockquote>
<div><p>[ Your answers here ]</p>
</div></blockquote>
</section>
</section>
<section id="plotting-the-family-tree">
<h3>2.2 Plotting the family tree<a class="headerlink" href="#plotting-the-family-tree" title="Permalink to this heading">#</a></h3>
<p>Printing the parent matrix for large simulations (with big populations and many generations) is not very helpful, because the network is too dense to properly visualise (you may try if you want).</p>
<p>Rather than looking the parent matrix, we will use the parent matrix to reconstruct a <em>family tree</em> for only the last generation. That is, we only look at the members of previous generations whose offspring appears in the last generation. We will then generate a visual representation of the tree.</p>
<section id="exercise-8-ancestors-in-a-tree">
<h4><strong>Exercise 8: Ancestors in a tree</strong><a class="headerlink" href="#exercise-8-ancestors-in-a-tree" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p><font color="orange"><b>Code:</b></font> In the code cell below, copy-paste the code for running the simulation with ancestry (the one above exercise 7, which generates the two plots). Now, set the <code class="docutils literal notranslate"><span class="pre">population_size</span></code> parameter to 100 and the <code class="docutils literal notranslate"><span class="pre">simulation_length</span></code> parameter to 1000, and run the simulation.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/7941c4ad1ae8e0c7ec32202a96cec380ce5572a4e48f0742faa21c3a505bf832.png" src="_images/7941c4ad1ae8e0c7ec32202a96cec380ce5572a4e48f0742faa21c3a505bf832.png" />
</div>
</div>
<p><font color="orange"><b>Code/Do:</b></font> Now, run the cell below to visualize the phylogenetic tree for the data you just generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">reconstruct_tree</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span>
<span class="nf">plot_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/d429f3ab21164a4052b58e3186f0be2455c40a5d0cc6098208e648efb7aa1b98.png" src="_images/d429f3ab21164a4052b58e3186f0be2455c40a5d0cc6098208e648efb7aa1b98.png" />
</div>
</div>
<p>Note: if the plot above is a bit hard to read, you can also generate a textual representation of the phylogenetic tree (this is called the ‘Newick string’) by running <code class="docutils literal notranslate"><span class="pre">print_tree(tree)</span></code> in the code cell below. From the result, copy everything between quotation marks (but exclude the two quotation marks themselves). Then, go to to <a class="reference external" href="https://icytree.org/">icytree.org</a> and click <em>File &gt; Enter tree directly…</em> Paste the textual representation you’ve just copied and click <em>Done</em>. Save the generated tree (<em>File &gt; Export tree as… &gt; PNG image</em>), and insert the image in the cell below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">print_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'(((((((84,93),(61)),((43,57),(((36),(21,64),(28,56,82,91)),(83),(14)))),(((18,60),(31,90)),((11,40),(20,45)))),((15),(47,55,69))),((((73,99),((27,38),(51))),((((22,70,87,89,92),((46),(58),(75))),(((67),(19)),(95,100))),((((32),(((59),((50),(6,62))),((8),(5)))),(49)),((37),(3))))),((((((33),(88)),(94)),((68,77),((79),(2)))),(((((16,24),(26)),((23,35,78),(9))),((29,71),((96),((98),(97))))),((((((44,53,66,76,86),(34,81)),((72),(85))),((25),(4,74))),((((10),(12,63)),((41,48),(7,39))),(80))),((54,65),(1),(30))))),((13),((17,52),(42)))))));'</div></div>
</div>
<blockquote>
<div><p>[ Your phylogenetic tree here ]</p>
</div></blockquote>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> As far as you can judge, how many generations ago did the LCA of the current population live?</p></li>
<li><p><font color="green"><b>Think:</b></font> Which aspects of evolution leave traces that we can detect in the current generation, and which aspects do not?</p></li>
</ul>
<blockquote>
<div><p>[ Your answers here ]</p>
</div></blockquote>
</section>
</section>
</section>
<section id="phylogenetic-reconstruction-of-biological-data">
<h2>3. Phylogenetic reconstruction of biological data<a class="headerlink" href="#phylogenetic-reconstruction-of-biological-data" title="Permalink to this heading">#</a></h2>
<p>In evolutionary research, the elaborate ancestry information represented by the parent matrix is usually not available. To reconstruct family trees we have to resort to different methods. Information about when species branched off (‘speciated’) can be deduced from genetic variation in the current population. For instance, horses are genetically more similar to donkeys than to, say, frogs. So, the last common ancestor of horses and frogs most likely lived much further in the past than the last common ancestor of horses and donkeys. In other words, the branch that would eventually evolve into frogs split off from the branch that would eventually evolve into horses earlier than the branch that would eventually evolve into donkeys. This type of analysis is called phylogenetic reconstruction. It’s based on genetic similarity between members of the current generation, which we measure using a <em>distance measure</em>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">phangorn</span></code> package (that we installed and loaded in the first code cell of this notebook) comes with a dataset that contains real genetic data (i.e., RNA samples) from many different species.</p>
<p>Run the following code cell to load that dataset:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">data</span><span class="p">(</span><span class="n">Laurasiatherian</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>With the code cell below, you can print a summary of the data:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">str</span><span class="p">(</span><span class="n">Laurasiatherian</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>List of 47
 $ Platypus  : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ Wallaroo  : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Possum    : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Bandicoot : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Opposum   : int [1:1605] 2 1 4 1 3 3 4 4 4 3 ...
 $ Armadillo : int [1:1605] 2 1 2 1 3 3 4 2 4 3 ...
 $ Elephant  : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Aardvark  : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ Tenrec    : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ Hedghog   : int [1:1605] 1 4 1 1 3 3 4 2 4 3 ...
 $ Gymnure   : int [1:1605] 1 1 1 1 3 3 4 2 4 3 ...
 $ Mole      : int [1:1605] 1 1 1 1 3 3 4 4 4 3 ...
 $ Shrew     : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ Rbat      : int [1:1605] 2 1 4 1 3 3 2 4 4 2 ...
 $ FlyingFox : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ RyFlyFox  : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ FruitBat  : int [1:1605] 4 1 4 1 3 3 4 4 4 3 ...
 $ LongTBat  : int [1:1605] 2 1 4 1 3 3 4 4 4 3 ...
 $ Horse     : int [1:1605] 2 1 4 1 3 3 2 4 4 3 ...
 $ Donkey    : int [1:1605] 2 1 4 1 3 3 2 4 4 3 ...
 $ WhiteRhino: int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ IndianRhin: int [1:1605] 2 1 4 1 3 3 4 4 4 3 ...
 $ Pig       : int [1:1605] 2 1 2 1 3 3 4 4 4 3 ...
 $ Alpaca    : int [1:1605] 2 1 2 1 3 3 4 4 4 1 ...
 $ Cow       : int [1:1605] 2 1 4 1 3 3 4 4 4 3 ...
 $ Sheep     : int [1:1605] 4 1 4 1 3 3 4 4 4 3 ...
 $ Hippo     : int [1:1605] 4 1 2 1 3 3 4 4 4 3 ...
 $ FinWhale  : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ BlueWhale : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ SpermWhale: int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Rabbit    : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ Pika      : int [1:1605] 1 1 4 1 3 3 4 4 4 3 ...
 $ Squirrel  : int [1:1605] 2 1 4 1 3 3 4 4 4 3 ...
 $ Dormouse  : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ GuineaPig : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Mouse     : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Vole      : int [1:1605] 1 1 1 1 3 3 4 4 4 3 ...
 $ CaneRat   : int [1:1605] 2 1 1 1 4 3 4 4 4 3 ...
 $ Baboon    : int [1:1605] 1 1 4 1 3 3 2 4 4 3 ...
 $ Human     : int [1:1605] 1 1 4 1 3 3 4 4 4 3 ...
 $ Loris     : int [1:1605] 2 1 2 1 3 3 4 4 4 3 ...
 $ Cebus     : int [1:1605] 2 1 2 1 3 3 4 4 4 3 ...
 $ Cat       : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ Dog       : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ HarbSeal  : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ FurSeal   : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 $ GraySeal  : int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 - attr(*, &quot;class&quot;)= chr &quot;phyDat&quot;
 - attr(*, &quot;weight&quot;)= int [1:1605] 1 5 1 444 3 314 1 1 328 1 ...
 - attr(*, &quot;nr&quot;)= int 1605
 - attr(*, &quot;nc&quot;)= num 4
 - attr(*, &quot;index&quot;)= int [1:3179] 1 2 3 4 5 6 7 8 9 6 ...
 - attr(*, &quot;levels&quot;)= chr [1:4] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot;
 - attr(*, &quot;allLevels&quot;)= chr [1:18] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot; ...
 - attr(*, &quot;type&quot;)= chr &quot;DNA&quot;
 - attr(*, &quot;contrast&quot;)= num [1:18, 1:4] 1 0 0 0 0 1 1 1 0 0 ...
  ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:4] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot;
</pre></div>
</div>
</div>
</div>
<p>The data originates from the (now closed) Allan Wilson Centre in New Zealand. To find out more about this data, have a look at <a class="reference external" href="http://www.allanwilsoncentre.ac.nz/massey/learning/departments/centres-research/allan-wilson-centre/our-research/resources/software-and-supplementary-materials.cfm">their website</a>.</p>
<p>We will try to reconstruct a phylogenetic tree for these species. That is, we will try to reconstruct when different species branched off from each other, based only on genetic information of the current population (the last generation). The first step is to measure ‘genetic distance’ between the genetic samples for each species. For simplicity, we assume that:</p>
<ol class="arabic simple">
<li><p>all species ultimately originate from a single common ancestor (an uncontroversial assumption in evolutionary biology);</p></li>
<li><p>species have diverged genetically by picking up mutations at a roughly constant rate (a more problematic assumption).</p></li>
</ol>
<p>(Try and convince yourself that the phylogenetic tree reconstruction method we described requires the second assumption — and that this assumption is problematic when considering evolution in the real world.)</p>
<p>The distance between strings of DNA or RNA is typically measured by counting the number of mutations required to change one into the other. Because of the second assumption, the genetic distance between two species is proportional to the time that has passed since their last common ancestor.</p>
<section id="exercise-9-animal-species-distances">
<h3><strong>Exercise 9: Animal species distances</strong><a class="headerlink" href="#exercise-9-animal-species-distances" title="Permalink to this heading">#</a></h3>
<p><font color="orange"><b>Code/Do:</b></font> Select five species from the Laurasiatherian dataset (for instance three that you think are closely related and two that are more distantly related). Create a subset of the data containing just these five species, by changing the numbers in the code cell below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mysubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">Laurasiatherian</span><span class="p">,</span><span class="w"> </span><span class="n">subset</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">19</span><span class="p">,</span><span class="m">20</span><span class="p">,</span><span class="m">28</span><span class="p">,</span><span class="m">29</span><span class="p">,</span><span class="m">30</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>The numbers correspond to the position of the species in the list printed by <code class="docutils literal notranslate"><span class="pre">str(Laurasiatherian)</span></code> above, i.e. Platypus = 1, Possum = 3, etc. ) You have to replace these numbers by the numbers corresponding to the species that you chose. Run the cell after you have done so. Then, verify that your subset contains the right species by running the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">str</span><span class="p">(</span><span class="n">mysubset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>List of 5
 $ Horse     : int [1:1605] 2 1 4 1 3 3 2 4 4 3 ...
 $ Donkey    : int [1:1605] 2 1 4 1 3 3 2 4 4 3 ...
 $ FinWhale  : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ BlueWhale : int [1:1605] 4 1 1 1 3 3 4 4 4 3 ...
 $ SpermWhale: int [1:1605] 2 1 1 1 3 3 4 4 4 3 ...
 - attr(*, &quot;class&quot;)= chr &quot;phyDat&quot;
 - attr(*, &quot;weight&quot;)= int [1:1605] 1 5 1 444 3 314 1 1 328 1 ...
 - attr(*, &quot;nr&quot;)= int 1605
 - attr(*, &quot;nc&quot;)= num 4
 - attr(*, &quot;index&quot;)= int [1:3179] 1 2 3 4 5 6 7 8 9 6 ...
 - attr(*, &quot;levels&quot;)= chr [1:4] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot;
 - attr(*, &quot;allLevels&quot;)= chr [1:18] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot; ...
 - attr(*, &quot;type&quot;)= chr &quot;DNA&quot;
 - attr(*, &quot;contrast&quot;)= num [1:18, 1:4] 1 0 0 0 0 1 1 1 0 0 ...
  ..- attr(*, &quot;dimnames&quot;)=List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:4] &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot;
</pre></div>
</div>
</div>
</div>
<p>Now, we compute the pairwise distance between all species in the subset using the function <code class="docutils literal notranslate"><span class="pre">dist.ml</span></code>. The results are stored in a <em>distance matrix</em>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">distance_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dist.ml</span><span class="p">(</span><span class="n">mysubset</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                Horse     Donkey   FinWhale  BlueWhale
Donkey     0.01814797                                 
FinWhale   0.11539023 0.11319215                      
BlueWhale  0.11392413 0.11173034 0.02657412           
SpermWhale 0.11502343 0.11796281 0.06811650 0.06777211
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> How can you read off the distance between two species from this matrix? Why are the numbers on the diagonal of this matrix zero?</p></li>
<li><p><font color="green"><b>Think:</b></font> Do the computed distances correspond to your intuitions about the selected species’ relatedness?</p></li>
</ul>
<blockquote>
<div><p>[ Your answers here ]</p>
</div></blockquote>
</section>
<section id="exercise-10-animal-species-phylogeny">
<h3><strong>Exercise 10: Animal species phylogeny</strong><a class="headerlink" href="#exercise-10-animal-species-phylogeny" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><font color="green"><b>Think/Do:</b></font> Using pen and paper, or your favourite drawing software, reconstruct a phylogenetic tree that describes the evolutionary relations between your selected species. Use the principles described earlier. You shouldn’t need to do any calculations.</p></li>
</ul>
<blockquote>
<div><p>[ Your hand-drawn phylogenetic tree here ]</p>
</div></blockquote>
</section>
</section>
<section id="phylogenetic-trees-with-hierarchical-clustering">
<h2>4. Phylogenetic trees with hierarchical clustering<a class="headerlink" href="#phylogenetic-trees-with-hierarchical-clustering" title="Permalink to this heading">#</a></h2>
<section id="upgma">
<h3>4.1 UPGMA<a class="headerlink" href="#upgma" title="Permalink to this heading">#</a></h3>
<p>We can use a simple method called ‘hierarchical clustering’ to build the phylogenetic trees such as in the previous sections automatically. Hierarchical clustering can be done with the UPGMA algorithm. (If you need a refresher on how exactly it worked, look up the UPGMA chapter from the readings again!) In short, UPGMA follows these steps:</p>
<ul class="simple">
<li><p>Treat each datapoint (for example, a RNA sample of a species) as a separate “cluster” containing just one datapoint;</p></li>
<li><p>Compute the distances between all clusters (using some distance measure; for example, genetic distance);</p></li>
<li><p>Merge the two clusters that are nearest to each other into a new cluster;</p></li>
<li><p>Repeat steps 2 and 3 until only all datapoints are in cluster.</p></li>
</ul>
</section>
<section id="constructing-a-tree">
<h3>4.2 Constructing a tree<a class="headerlink" href="#constructing-a-tree" title="Permalink to this heading">#</a></h3>
<p>To construct a phylogenetic tree, we can think of each merging of clusters as the joining of two branches. In the simplest version of this algorithm, we define ‘distance’ between a cluster A and a cluster B as the average distance between any datapoint in A and any datapoint in B — this is what UPGMA does. A slightly more complicated method, Ward’s clustering, uses the square root of the average of the squared point-to-point distances.</p>
<section id="exercise-11-animal-species-phylogeny-with-upgma">
<h4><strong>Exercise 11: Animal species phylogeny, with UPGMA</strong><a class="headerlink" href="#exercise-11-animal-species-phylogeny-with-upgma" title="Permalink to this heading">#</a></h4>
<ul class="simple">
<li><p><font color="green"><b>Think/Do:</b></font> Using the distances between species in <code class="docutils literal notranslate"><span class="pre">mysubset</span></code> (from the distance matrix you computed in the previous section), manually perform three cycles of the UPGMA algorithm with pen and paper.</p></li>
</ul>
<blockquote>
<div><p>[ Your UPGMA steps here ]</p>
</div></blockquote>
</section>
<section id="exercise-12-animal-phylogenetic-tree-from-upgma">
<h4><strong>Exercise 12: Animal phylogenetic tree, from UPGMA</strong><a class="headerlink" href="#exercise-12-animal-phylogenetic-tree-from-upgma" title="Permalink to this heading">#</a></h4>
<p>Generate a phylogenetic tree for your subset and plot it using the code cell below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">upgma</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s">&#39;average&#39;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">0.65</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/1e44d99aaa91fede6829ca05ac0c6dbad63014e87a498d0c446b9b310e60313a.png" src="_images/1e44d99aaa91fede6829ca05ac0c6dbad63014e87a498d0c446b9b310e60313a.png" />
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> Is the tree the same as the one that you created before with pen and paper?</p></li>
<li><p><font color="orange"><b>Code:</b></font> In the code cell below, create a tree for the entire dataset. To do this, you first need to compute the distance matrix for the entire dataset, and then run the upgma algorithm on it.</p></li>
<li><p><font color="green"><b>Think:</b></font> Does the generated tree agree with your expectations?</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ee7df16d2aba9a5e3490717c36b61c4190d0eae11e1dfe3e256ba79a86fe67a2.png" src="_images/ee7df16d2aba9a5e3490717c36b61c4190d0eae11e1dfe3e256ba79a86fe67a2.png" />
</div>
</div>
<blockquote>
<div><p>[ Your answers here ]</p>
</div></blockquote>
<p><em>Optional:</em> Try different methods for computing the distance between clusters by changing the parameter <code class="docutils literal notranslate"><span class="pre">method</span></code> from <code class="docutils literal notranslate"><span class="pre">'average'</span></code> to something else (options are, for instance, <code class="docutils literal notranslate"><span class="pre">'ward.D'</span></code>, <code class="docutils literal notranslate"><span class="pre">'single'</span></code> and <code class="docutils literal notranslate"><span class="pre">'median'</span></code>). Do you notice any changes in the resulting phylogenetic trees?</p>
</section>
<section id="exercise-13-our-simulated-ancestors-in-a-tree-with-upgma">
<h4><strong>Exercise 13: Our simulated ancestors in a tree, with UPGMA</strong><a class="headerlink" href="#exercise-13-our-simulated-ancestors-in-a-tree-with-upgma" title="Permalink to this heading">#</a></h4>
<p>We will now investigate what happens if we perform phylogenetic analysis on the population resulting from our own simulated evolution. Remember that, since this is a simulation over which we have full control, we can reconstructed the actual phylogenetic tree using the information that we stored in the parent matrix.</p>
<p>Generate a distance matrix between the members of the last generation from your simulation, and reconstruct a phylogenetic tree with the <code class="docutils literal notranslate"><span class="pre">upgma</span></code> function (choose your preferred <em>method</em>, e.g. <code class="docutils literal notranslate"><span class="pre">ward.D</span></code>):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">distance_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_distance_matrix</span><span class="p">(</span><span class="n">population</span><span class="p">)</span>
<span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">upgma</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s">&#39;ward.D&#39;</span><span class="p">)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">0.4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/b43ef6047f4d932d944f4d310686632fe603a9271ac27a96ad9e9c23733377d6.png" src="_images/b43ef6047f4d932d944f4d310686632fe603a9271ac27a96ad9e9c23733377d6.png" />
</div>
</div>
<p>If the tree above is a bit hard to read again, use the cell below to print the Newick string that you can paste into <a class="reference external" href="https://icytree.org/">icytree.org</a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">write.tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'(((15:0.03,(69:0,(47:0,55:0):0):0.03):0.5004195804,(((84:0.01666666667,(61:0.01,93:0.01):0.006666666667):0.08166666667,(60:0.01785714286,(40:0.01714285714,(90:0,(45:0,(31:0,(20:0,(11:0,18:0):0):0):0):0):0.01714285714):0.0007142857143):0.08047619048):0.1034848485,((43:0,57:0):0.1054545455,((91:0,(82:0,(28:0,56:0):0):0):0.048,(83:0.017,(36:0.015,(64:0,(14:0,21:0):0):0.015):0.002):0.031):0.05745454545):0.09636363636):0.3286013986):3.889839463,(((32:0.03688888889,((5:0.01,8:0.01):0.01396825397,(37:0.01714285714,(62:0,(59:0,(50:0,(49:0,(3:0,6:0):0):0):0):0):0.01714285714):0.006825396825):0.01292063492):0.4344139434,(((73:0,99:0):0.05333333333,(27:0.02666666667,(38:0,51:0):0.02666666667):0.02666666667):0.3608823529,(((19:0.01,67:0.01):0.04,(95:0.01,100:0.01):0.04):0.2041666667,((92:0,(89:0,(87:0,(22:0,70:0):0):0):0):0.05416666667,(75:0.03333333333,(46:0.02,58:0.02):0.01333333333):0.02083333333):0.2):0.1600490196):0.05708714597):1.075564887,((((41:0.01,48:0.01):0.02916666667,(80:0.01733333333,(10:0.016,(63:0,(39:0,(7:0,12:0):0):0):0.016):0.001333333333):0.02183333333):0.09541666667,((65:0.01833333333,(30:0.01666666667,(1:0.01,54:0.01):0.006666666667):0.001666666667):0.06125,(25:0.05575757576,((4:0.02,74:0.02):0.03313131313,((72:0.01,85:0.01):0.03206349206,((86:0,(76:0,(66:0,(44:0,53:0):0):0):0):0.03571428571,(34:0.01,81:0.01):0.02571428571):0.006349206349):0.01106782107):0.002626262626):0.02382575758):0.055):0.4232442954,(((24:0.01714285714,(78:0,(35:0,(26:0,(23:0,(9:0,16:0):0):0):0):0):0.01714285714):0.09604761905,(98:0.018,(97:0.01666666667,(96:0.01333333333,(29:0,71:0):0.01333333333):0.003333333333):0.001333333333):0.09519047619):0.2461837004,((13:0.05166666667,(42:0.01333333333,(17:0,52:0):0.01333333333):0.03833333333):0.2074891775,((94:0.02666666667,(33:0,88:0):0.02666666667):0.08523809524,((68:0,77:0):0.03,(2:0.02,79:0.02):0.01):0.0819047619):0.1472510823):0.1002183324):0.1984534522):0.98904009):2.873391325);'</div></div>
</div>
<p>Now generate the <em>actual</em> family tree of the simulation by running the cell below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">gold_standard_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">reconstruct_tree</span><span class="p">(</span><span class="n">parent_matrix</span><span class="p">)</span>
<span class="nf">plot_tree</span><span class="p">(</span><span class="n">gold_standard_tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/d429f3ab21164a4052b58e3186f0be2455c40a5d0cc6098208e648efb7aa1b98.png" src="_images/d429f3ab21164a4052b58e3186f0be2455c40a5d0cc6098208e648efb7aa1b98.png" />
</div>
</div>
<p>Newick string for pasting into <a class="reference external" href="https://icytree.org/">icytree.org</a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">print_tree</span><span class="p">(</span><span class="n">gold_standard_tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">'(((((((84,93),(61)),((43,57),(((36),(21,64),(28,56,82,91)),(83),(14)))),(((18,60),(31,90)),((11,40),(20,45)))),((15),(47,55,69))),((((73,99),((27,38),(51))),((((22,70,87,89,92),((46),(58),(75))),(((67),(19)),(95,100))),((((32),(((59),((50),(6,62))),((8),(5)))),(49)),((37),(3))))),((((((33),(88)),(94)),((68,77),((79),(2)))),(((((16,24),(26)),((23,35,78),(9))),((29,71),((96),((98),(97))))),((((((44,53,66,76,86),(34,81)),((72),(85))),((25),(4,74))),((((10),(12,63)),((41,48),(7,39))),(80))),((54,65),(1),(30))))),((13),((17,52),(42)))))));'</div></div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> How well does the reconstruction produced by the hierarchical clustering algorithm match the actual family tree?</p></li>
<li><p><font color="green"><b>Think:</b></font> How can you explain the differences between the reconstructed and the actual family tree?</p></li>
</ul>
</section>
</section>
</section>
<section id="phylogenetic-reconstruction-of-languages">
<h2>5. Phylogenetic reconstruction of languages<a class="headerlink" href="#phylogenetic-reconstruction-of-languages" title="Permalink to this heading">#</a></h2>
<p>In the tutorial you have learned about features of language and music used in comparative research, and about variation across languages and musics. By studying some specific examples, you have seen that both languages and musical traditions are transmitted culturally, and are subject to a process of cultural evolution. In this final section of the computer lab, you will use the methods that we saw for phylogenetic tree reconstruction in the previous sections to reconstruct the cultural evolutionary history of languages.</p>
<section id="exercise-14-languages-in-a-tree-with-upgma">
<h3><strong>Exercise 14: Languages in a tree, with UPGMA</strong><a class="headerlink" href="#exercise-14-languages-in-a-tree-with-upgma" title="Permalink to this heading">#</a></h3>
<p>We preprocessed a dataset for you so it can be loaded into R. To do this, run the following code cell:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">load</span><span class="p">(</span><span class="s">&quot;data/language_data.Rdata&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This will create an object called <code class="docutils literal notranslate"><span class="pre">mydata</span></code> containing the dataset. You can check the languages in the data as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">names</span><span class="p">(</span><span class="n">mydata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class=list-inline><li>'Hittite'</li><li>'Luvian'</li><li>'Lycian'</li><li>'Tocharian_A'</li><li>'Tocharian_B'</li><li>'Albanian_C'</li><li>'Albanian_K'</li><li>'Albanian_G'</li><li>'Albanian_Top'</li><li>'Classical_Armenian'</li><li>'Armenian_List'</li><li>'Armenian_Mod'</li><li>'Ancient_Greek'</li><li>'Avestan'</li><li>'Old_Persian'</li><li>'Persian_List'</li><li>'Afghan'</li><li>'Wakhi'</li><li>'Baluchi'</li><li>'Waziri'</li><li>'Tadzik'</li><li>'Kurdish'</li><li>'Iron_Ossetic'</li><li>'Digor_Ossetic'</li><li>'Greek_Mod'</li><li>'Greek_ML'</li><li>'Vedic_Sanskrit'</li><li>'Nepali_List'</li><li>'Assamese'</li><li>'Oriya'</li><li>'Bengali'</li><li>'Bihari'</li><li>'Marwari'</li><li>'Hindi'</li><li>'Urdu'</li><li>'Sindhi'</li><li>'Lahnda'</li><li>'Gujarati'</li><li>'Marathi'</li><li>'Kashmiri'</li><li>'Singhalese'</li><li>'Romani'</li><li>'Serbocroatian'</li><li>'Bulgarian'</li><li>'Macedonian'</li><li>'Russian'</li><li>'Polish'</li><li>'Byelorussian'</li><li>'Ukrainian'</li><li>'Czech_E'</li><li>'Slovak'</li><li>'Czech'</li><li>'Lusatian_U'</li><li>'Lusatian_L'</li><li>'Slovenian'</li><li>'Old_Church_Slavonic'</li><li>'Old_Prussian'</li><li>'Latvian'</li><li>'Lithuanian_ST'</li><li>'Faroese'</li><li>'Icelandic_ST'</li><li>'Riksmal'</li><li>'Danish'</li><li>'Swedish_List'</li><li>'Swedish_VL'</li><li>'Swedish_Up'</li><li>'Old_Norse'</li><li>'Old_High_German'</li><li>'Old_English'</li><li>'English_ST'</li><li>'Frisian'</li><li>'Flemish'</li><li>'Dutch_List'</li><li>'Luxembourgish'</li><li>'German_ST'</li><li>'Gothic'</li><li>'Catalan'</li><li>'Portuguese_ST'</li><li>'Spanish'</li><li>'Sardinian_C'</li><li>'Sardinian_L'</li><li>'Sardinian_N'</li><li>'Walloon'</li><li>'French'</li><li>'Provencal'</li><li>'Ladin'</li><li>'Romansh'</li><li>'Friulian'</li><li>'Italian'</li><li>'Romanian_List'</li><li>'Vlach'</li><li>'Latin'</li><li>'Oscan'</li><li>'Umbrian'</li><li>'Old_Irish'</li><li>'Cornish'</li><li>'Breton_ST'</li><li>'Breton_SE'</li><li>'Breton_List'</li><li>'Welsh_C'</li><li>'Welsh_N'</li><li>'Irish_A'</li><li>'Scots_Gaelic'</li></ol>
</div></div>
</div>
<ul class="simple">
<li><p><font color="orange"><b>Code/Do:</b></font> Choose a subset of the list of languages. We will initially build a phylogenetic tree of this subset. Define your subset with the subset function. For instance, if you want to select languages 13, 14, 15, 63 and 71 you type:</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mysubset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">mydata</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">13</span><span class="o">:</span><span class="m">15</span><span class="p">,</span><span class="m">63</span><span class="p">,</span><span class="m">71</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>Create a distance matrix of your subset, by letting the computer count the number of feature values that differ between two languages (“hamming distance”):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">distance_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dist.hamming</span><span class="p">(</span><span class="n">mysubset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Pick your favourite clustering algorithm and method and generate a tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">upgma</span><span class="p">(</span><span class="n">distance_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s">&#39;ward.D&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Plot your tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">use.edge.length</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/ca5ff80cbae3a27149e4796888627864df1b713f8fc4aad7ff5df3a7b4307990.png" src="_images/ca5ff80cbae3a27149e4796888627864df1b713f8fc4aad7ff5df3a7b4307990.png" />
</div>
</div>
<ul class="simple">
<li><p><font color="orange"><b>Code:</b></font> In the code cell below, do the same thing for the entire dataset (i.e. compute the distance matrix, run UPGMA, plot the tree). You might want to adapt the <code class="docutils literal notranslate"><span class="pre">cex</span></code> parameter in the plot function, that sets the font size. Be aware of the influence the clustering algorithm and method for computing the distances between clusters can have.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### YOUR CODE HERE ###</span>
</pre></div>
</div>
</div>
</div>
<p>After generating the tree above, you can use the following again to get the Newick string to paste into <a class="reference external" href="https://icytree.org/">icytree.org</a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">write.tree</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<ul class="simple">
<li><p><font color="green"><b>Think:</b></font> What are the nine main language families you can distinguish within the Indo-European family, and in which regions of the world are they spoken (before colonial times)?</p></li>
</ul>
<blockquote>
<div><p>[ Your answer here ]</p>
</div></blockquote>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="Bioacoustics_in_Praat.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bioacoustics in Praat</p>
      </div>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#getting-started-with-r">0. Getting started with R</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#installing-packages">Installing packages</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-dataset">Downloading dataset</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#auxiliary-functions">Auxiliary functions</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulated-evolution">1. Simulated evolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#creating-a-population">1.1 Creating a population</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1-creating-a-genotype"><strong>Exercise 1: Creating a genotype</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2-creating-the-population"><strong>Exercise 2: Creating the population</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-a-fitness-function">1.2 Defining a fitness function</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3-compute-each-member-s-fitness-value"><strong>Exercise 3: Compute each member’s fitness value</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#implementing-selection">1.3 Implementing selection</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-4-a-new-population"><strong>Exercise 4: A new population</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-5-population-fitness-over-generations"><strong>Exercise 5: Population fitness over generations</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evolution-with-mutation">1.4 Evolution with mutation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-6-mu-tating-genomes"><strong>Exercise 6: Mu-tating genomes</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#simulated-evolution-with-ancestry">2. Simulated evolution with ancestry</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tracking-ancestry">2.1 Tracking ancestry</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-7-finding-the-ancestors"><strong>Exercise 7: Finding the ancestors</strong></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-family-tree">2.2 Plotting the family tree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-8-ancestors-in-a-tree"><strong>Exercise 8: Ancestors in a tree</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-reconstruction-of-biological-data">3. Phylogenetic reconstruction of biological data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-9-animal-species-distances"><strong>Exercise 9: Animal species distances</strong></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-10-animal-species-phylogeny"><strong>Exercise 10: Animal species phylogeny</strong></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-trees-with-hierarchical-clustering">4. Phylogenetic trees with hierarchical clustering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#upgma">4.1 UPGMA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constructing-a-tree">4.2 Constructing a tree</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-11-animal-species-phylogeny-with-upgma"><strong>Exercise 11: Animal species phylogeny, with UPGMA</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-12-animal-phylogenetic-tree-from-upgma"><strong>Exercise 12: Animal phylogenetic tree, from UPGMA</strong></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-13-our-simulated-ancestors-in-a-tree-with-upgma"><strong>Exercise 13: Our simulated ancestors in a tree, with UPGMA</strong></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phylogenetic-reconstruction-of-languages">5. Phylogenetic reconstruction of languages</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-14-languages-in-a-tree-with-upgma"><strong>Exercise 14: Languages in a tree, with UPGMA</strong></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="https://licensebuttons.net/l/by-nc-sa/3.0/88x31.png"></a>
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>